inc/cropper/namespace.php

```diff
@@ -488,3 +510,374 @@ function prevent_thumbnail_generation( $sizes ) {
+		if ( empty( $height ) ) {
+			list( $width, $height ) = wp_constrain_dimensions( $image_meta['width'], $image_meta['height'], $width );
+		}
+	} else {
+		if ( isset( $tachyon_args['w'] ) ) {
+			$width = (int) $tachyon_args['w'];
+		}
+		if ( isset( $tachyon_args['h'] ) ) {
+			$height = (int) $tachyon_args['h'];
+		}
+		if ( $width && ! $height ) {
+			list( $width, $height ) = wp_constrain_dimensions( $image_meta['width'], $image_meta['height'], $width );
+		}
+		if ( ! $width && $height ) {
+			list( $width, $height ) = wp_constrain_dimensions( $image_meta['width'], $image_meta['height'], 0, $height );
+		}
+	}
+
+	// Still stumped?
+	if ( ! $width || ! $height ) {
+		return $image;
+	}
+
+	$size_array = [ $width, $height ];
+	$srcset = wp_calculate_image_srcset( $size_array, $image_src, $image_meta, $attachment_id );
+
+	if ( $srcset ) {
+		// Check if there is already a 'sizes' attribute.
+		$sizes = strpos( $image, ' sizes=' );
+		if ( ! $sizes ) {
+			$sizes = wp_calculate_image_sizes( $size_array, $image_src, $image_meta, $attachment_id );
+		}
+	}
+
+	if ( $srcset && $sizes ) {
+		// Format the 'srcset' and 'sizes' string and escape attributes.
+		$attr = sprintf( ' srcset="%s"', esc_attr( $srcset ) );
+		if ( is_string( $sizes ) ) {
+			$attr .= sprintf( ' sizes="%s"', esc_attr( $sizes ) );
+		}
+		// Add 'srcset' and 'sizes' attributes to the image markup.
+		$image = preg_replace( '/<img ([^>]+?)[\/ ]*>/', '<img $1' . $attr . ' />', $image );
+	}
+
+	return $image;
+}
+
+/**
+ * Return a list of modifiers for calculating image srcset and sizes from.
+ *
+ * @return array
+ */
+function get_image_size_modifiers( ?int $attachment_id = null ) : array {
+	/**
+	 * Filters the default image size modifiers. By default
+	 * the srcset will contain a 2x, 1.5x, 0.5x and 0.25x version of the image.
+	 *
+	 * @param array $modifiers The zoom values for the srcset.
+	 * @param int? $attachment_id The attachment ID or null.
+	 */
+	return apply_filters( 'hm.smart-media.image-size-modifiers', [ 2, 1.5, 0.5, 0.25 ], $attachment_id );
+}
+
+/**
+ * Update the sources array to return tachyon URLs that respect
+ * requested image aspect ratio and crop data.
+ *
+ * @param array   $sources       Array of source URLs and widths to generate the srcset attribute from.
+ * @param array   $size_array    Width and height of the original image.
+ * @param string  $image_src     The requested image URL.
+ * @param array   $image_meta    The image meta data array.
+ * @param integer $attachment_id The image ID.
+ * @return array
+ */
+function image_srcset( array $sources, array $size_array, string $image_src, array $image_meta, int $attachment_id ) : array {
+	$width = absint( $size_array[0] );
+	$height = absint( $size_array[1] );
+
+	if ( strpos( $image_src, TACHYON_URL ) === false ) {
+		$image_src = tachyon_url( $image_src );
+	}
+
+	// Multipliers for output srcset.
+	$modifiers = get_image_size_modifiers( $attachment_id );
```

Comment:
Is it feasible to start with the target widths and calculate the zoom factor from there? 

Seems like this is the primary config surface exposed to theme developers, and as a theme developer, it would feel more natural to specify a list of target widths than a list of zoom ratios.

Reply:
Yeah I am curious now, on our projects do you think we've taken the target width approach because of how WP outputs image `srcset` & `sizes` normally?

I thought to keep it simple like this because the default output of `sizes` attribute specifies the maximum width of the image which is why I thought 2x is the most you'd ever want or need, plus that's 2x etc the defined image size so presumably it's all the optimisation needed for the design. Happy to be corrected there!

I like your thinking here and I have some questions:

- are target widths universal
- should they affect the sizes attribute too eg. from twentynineteen featured images: `sizes="(max-width: 34.9rem) calc(100vw - 2rem), (max-width: 53rem) calc(8 * (100vw / 12)), (min-width: 53rem) calc(6 * (100vw / 12)), 100vw"`
- would a wrapper function be useful to generate image markup with custom target widths

Reply:
I'm thinking that as a theme developer, you want to define a couple basic things about how images will be handled: your crop ratios, the sizes attributes for images in various locations, and then the target widths that intermediate sizes should be generated at for srcsets.

The target widths is tightly dependant on the sizes definition, but it shouldn't also be coupled to the crop dimensions. For example, using Tachyon images, you should be able define your square crop size as 20x20 and it shouldn't affect the size that srcsets are displayed at.

> are target widths universal

No, not really. I imagine it's going to be really hard to pass the right context into a a filter that generates the target widths, but maybe passing in the image crop size requested, any "sizes" defined, and something about the current call stack (at the very least, whether this is happening in `the_content` filter) would be enough to select the right values.

> should they affect the sizes attribute too

When a theme calls one of the `wp_get_attachment_image()` family of functions directly, it can pass its own "sizes" definition.  And it's going to be hard to filter this correctly within the content. So I don't think we need to handle that in the definition of the target widths. 

Perhaps passing the "sizes" attribute, if defined, into the filter around the target widths, would give enough content for theme functionality to return the appropriate target widths depending on the sizes attribute.

> would a wrapper function be useful to generate image markup with custom target widths

Maybe! Though, as I mentioned above, `wp_get_attachment_image()` already works quite well for that, so long as there was a way to filter the srcset array.